feat: datanode decentralized history (#1336)
vegaprotocol/specs: protocol/0076-DANO-data-node.md

### Data integrity
1. Data produced in the core snapshots aligns with the data-node data proving that what is returned by data-node matches core state at any given block height (<a name="0076-DANO-004" href="#0076-DANO-004">0076-DANO-004</a>)

### Data-node decentralized history:

> 1.  Historical data must be available to load into the datanode and must not be dependent on any centralized entity.

Reviewer: ```suggestion
1.  Historical data must be available to load into the datanode and must not be dependent on any centralized entity. (<a name="0076-DANO-006" href="#0076-DANO-005">0076-DANO-005</a>)
```

> 1.  A datanode restored from decentralised history for a given block span must match exactly the state of a datanode that has the same block span of data created by consuming events.   

Reviewer: ```suggestion
1.  A datanode restored from decentralised history for a given block span must match exactly the state of a datanode that has the same block span of data created by consuming events. (<a name="0076-DANO-012" href="#0076-DANO-012">0076-DANO-012</a>)
```

1.  As the network produces more blocks the data should be stored correctly in the data-node after a data-node is restored from decentralized history. For example: Start a data-node from a given history segment for a known block height, ensure the datanode continues to update from that block onwards. (<a name="0076-DANO-007" href="#0076-DANO-007">0076-DANO-007</a>).  

> 1.  It should not be necessary to restore the full history (i.e. from genesis block) to be able to process new blocks.  Restoring just the most recent history segment should be sufficient or the node to process new blocks.

Reviewer: ```suggestion
1.  It should not be necessary to restore the full history (i.e. from genesis block) to be able to process new blocks.  Restoring just the most recent history segment should be sufficient or the node to process new blocks. (<a name="0076-DANO-015" href="#0076-DANO-015">0076-DANO-015</a>)
```

1.  No data is duplicated as the core emits events when catching up to the later block height. For example: Starting a core node at block height less than the data-node block height must result in no duplicated data (<a name="0076-DANO-008" href="#0076-DANO-008">0076-DANO-008</a>)

> 1.  Starting a core node at block height greater than the data-nodes block height must result in an error an a refusal to start (<a name="0076-DANO-014" href="#0076-DANO-014">0076-DANO-014</a>)

Reviewer 2: ```suggestion
1.  Starting a core node at block height greater than the data-nodes block height must result in an error and a refusal to start (<a name="0076-DANO-014" href="#0076-DANO-014">0076-DANO-014</a>)
```

> 1.  If a data-node snapshot fails during the restore the process, it should error and the node(s) wont start (<a name="0076-DANO-009" href="#0076-DANO-009">0076-DANO-009</a>)

Reviewer 2: ```suggestion
1.  If a data-node snapshot fails during the restore the process, it should error and the node(s) won't start (<a name="0076-DANO-009" href="#0076-DANO-009">0076-DANO-009</a>)
```

> 1.   When queried via the APIs a node restored from decentralized history should return identical results to a node with the same block span which has been populated by event consumption.  [project front end dApps](https://github.com/vegaprotocol/frontend-monorepo/actions/workflows/generate-queries.yml). (<a name="0076-COSMICELEVATOR-006" href="#0076-COSMICELEVATOR-006">0076-COSMICELEVATOR-006</a>)

Reviewer 2: ```suggestion
1.   When queried via the APIs a node restored from decentralized history should return identical results to a node with the same block span which has been populated by event consumption.  [project front end dApps](https://github.com/vegaprotocol/frontend-monorepo/actions/workflows/generate-queries.yml). (<a name="0076-DANO-014" href="#0076-DANO-014">0076-DANO-014</a>)
```
